spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -329,8 +329,8 @@ private SendingHandler(MessageHandler delegate, boolean embedHeaders,
 
 		@Override
 		protected void handleMessageInternal(Message<?> message) throws Exception {
-			Message<?> messageToSend = (this.useNativeEncoding) ? getMessageBuilderFactory().withPayload(message.getPayload()).build() :
-					serializeAndEmbedHeadersIfApplicable(message);
+			Message<?> messageToSend = (this.useNativeEncoding) ? getMessageBuilderFactory().withPayload(message.getPayload())
+					.copyHeaders(message.getHeaders()).build() : serializeAndEmbedHeadersIfApplicable(message);
```

Comment:
We can just return the original message here and skip the redundant message re-creation. Will fix on merge.